sql/catalyst/src/main/scala/org/apache/spark/sql/util/ToJsonUtil.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.util
+
+import java.sql.{Date, Timestamp}
+import java.time.{Instant, LocalDate}
+import java.util.Base64
+
+import scala.collection.mutable
+
+import org.json4s._
+import org.json4s.JsonAST.JValue
+import org.json4s.jackson.JsonMethods.{compact, pretty, render}
+
+import org.apache.spark.annotation.Unstable
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.CatalystTypeConverters
+import org.apache.spark.sql.catalyst.util.{DateFormatter, DateTimeUtils, TimestampFormatter}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types._
+import org.apache.spark.unsafe.types.CalendarInterval
+
+
+object ToJsonUtil {
+  /**
+   * The compact JSON representation of this row.
+   * @since 3.5
+   */
+  @Unstable
+  def json(row: Row): String = compact(jsonValue(row))
+
+  /**
+   * The pretty (i.e. indented) JSON representation of this row.
+   * @since 3.5
+   */
+  @Unstable
+  def prettyJson(row: Row): String = pretty(render(jsonValue(row)))
+
+  /**
+   * JSON representation of the row.
+   *
+   * Note that this only supports the data types that are also supported by
+   * [[org.apache.spark.sql.catalyst.encoders.RowEncoder]].
+   *
+   * @return the JSON representation of the row.
+   */
+  private[sql] def jsonValue(row: Row): JValue = {
+    require(row.schema != null, "JSON serialization requires a non-null schema.")
+
+    lazy val zoneId = DateTimeUtils.getZoneId(SQLConf.get.sessionLocalTimeZone)
+    lazy val dateFormatter = DateFormatter()
+    lazy val timestampFormatter = TimestampFormatter(zoneId)
+
+    // Convert an iterator of values to a json array
+    def iteratorToJsonArray(iterator: Iterator[_], elementType: DataType): JArray = {
+      JArray(iterator.map(toJson(_, elementType)).toList)
+    }
+
+    // Convert a value to json.
+    def toJson(value: Any, dataType: DataType): JValue = (value, dataType) match {
```

Comment:
should we make this function to `private[sql]`, `JValue` is  a definition in json4s, and we should try to avoid exposing it to users in the public api

Reply:
Perhaps we can use the Jackson API to rewrite this logic to weaken our dependence on json4s, which could be another pr.

Reply:
Yeah, it's my bad

Reply:
> Perhaps we can use the Jackson API to rewrite this logic to weaken our dependence on json4s, which could be another pr (https://issues.apache.org/jira/browse/SPARK-39658 recorded this, but did not start discussing).

+1